numpy/core/src/multiarray/compiled_base.c

```diff
@@ -80,6 +80,358 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+            }
+        }
+        NPY_END_ALLOW_THREADS;
+
+        Py_DECREF(arr_w);
+    }
+
+    Py_DECREF(arr_edges);
+    Py_DECREF(arr_x);
+    return (PyObject *)hist;
+
+fail:
+    Py_XDECREF(hist);
+    Py_XDECREF(arr_w);
+    Py_XDECREF(arr_edges);
+    Py_XDECREF(arr_x);
+    return NULL;
+}
+
+NPY_NO_EXPORT PyObject *
+arr_histogramdd_uniform(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
+{
+    PyObject *obj_x, *obj_bins, *obj_edges, *weights = Py_None;
+    PyArrayObject *arr_x = NULL, *arr_bins = NULL, *arr_edges = NULL,
+                  *arr_w = NULL, *hist = NULL;
+    npy_intp *bins, len, ndim, d, i, ii, ix[NPY_MAXDIMS], nedges_total;
+    double *x = NULL, *w = NULL, min[NPY_MAXDIMS], max[NPY_MAXDIMS],
+           norm[NPY_MAXDIMS], tx;
+    int overflow;
+
+    static char *kwlist[] = {"x", "nbins", "edges", "weights", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OOO|O:_histogramdd_uniform",
+            kwlist, &obj_x, &obj_bins, &obj_edges, &weights)) {
+        goto fail;
+    }
+
+    /*
+     * Get data (x) which is of shape (len, ndim)
+     * where len is the number of samples
+     * and ndim is the number of dimensions
+     */
+    arr_x = (PyArrayObject *)PyArray_ContiguousFromAny(obj_x, NPY_DOUBLE, 2, 2);
+    if (arr_x == NULL) {
+        goto fail;
+    }
+    len = PyArray_DIM(arr_x, 0);
+    ndim = PyArray_DIM(arr_x, 1);
+
+    /*
+     * Get the number of bins for each array since
+     * the edges are passed in as a flat array
+     */
+    arr_bins = (PyArrayObject *)PyArray_ContiguousFromAny(obj_bins, NPY_INTP, 1, 1);
+    if (arr_bins == NULL) {
+        goto fail;
+    }
+    if (PyArray_DIM(arr_bins, 0) != ndim) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Number of bins array must be same length as array");
+        }
+        goto fail;
+    }
+    bins = (npy_intp *)PyArray_DATA(arr_bins);
+
+    /* Handle empty array */
+    if (len == 0) {
+        hist = (PyArrayObject *)PyArray_ZEROS(ndim, bins, NPY_INTP, 0);
+        if (hist == NULL) {
+            goto fail;
+        }
+        Py_DECREF(arr_bins);
+        Py_DECREF(arr_x);
+        return (PyObject *)hist;
+    }
+
+    /*
+     * Get bin edges which is a flat array of length sum(nbins)
+     */
+    arr_edges = (PyArrayObject *)PyArray_ContiguousFromAny(obj_edges, NPY_DOUBLE, 1, 1);
+    if (arr_edges == NULL) {
+        goto fail;
+    }
+    nedges_total = 0;
+    for (i = 0; i < ndim; i++) {
+        nedges_total += bins[i] + 1;
+    }
+    if (PyArray_SIZE(arr_edges) != nedges_total) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Length of edges must be equal to the total number of edges");
+        }
+        goto fail;
+    }
+
+    x = (double *)PyArray_DATA(arr_x);
+
+    {
+        double *e = (double *)PyArray_DATA(arr_edges);
+        for (d = 0; d < ndim; d++) {
```

Comment:
Why not move the declaration of `d` inside this block?

Reply:
I was following the style established in bincount just below these histogram functions where all variables are declared up top. I would be more than happy to move declarations to the tightest scope if numpy's coding style has changed since arr_bincount() was written.

Reply:
Everything declared at the top is a bad style still present in old code, you do not have to stick to that. Declaring in the smallest scope is usually much cleaner.